### ISSUE TITLE
ONOS 1.9 disconnects from switch (OF 1.3)

### ISSUE THREAD

**Issue by vkochan** [NorthboundNetworks/ZodiacFX#84]
Hi Zodiac community,

I observed an issue when ONOS 1.9 disconnects from the switch probably in case of
multi-part messages. I tried to look ta the code and found probably theoretical place of issue (but I did not test it):

in below function "multi_pos" variable is checked to send multi-part message:

void of13_message(struct ofp_header *ofph, int size, int len) {
    ...
    if (size == len && multi_pos !=0)
	{
		sendtcp(&shared_buffer, multi_pos);
	}
	return;

but it is cleared only at:
of_receive(...) {
    ...
    while (size < len)
		{
			ofph = &packetbuffer[size];
			if (size == 0) multi_pos = 0;

might be there an issue when we received 2 messages by 1 read from tcp - 1st multi-part and some-other ? if so then I suspect that there might be issue when 2nd message is handled the following case will be triggered:

       if (size == len && multi_pos !=0)
	{
		sendtcp(&shared_buffer, multi_pos);
	}

What do you think ?

Thanks,

**Comment by mwutzke** [comment]
I've seen similar issues as well (when using Faucet), resulting not just in disconnects but also inconsistent flows / tables. As indicated by @vkochan, I suspect there are issues with the processing mutiple OF messages in 1 call to of_receive(), esp where 1 OF message spans multiple TCP segments (TCP is a byte steam procotol, and does not preserve sender datagram boundries like UDP does).

I've been able to simpulate this operation by introducing a 50ms delay in the OF controller - ZodiacFX data path with something like the following:

```
sudo tc qdisc add dev $DVETH root netem delay 10ms
```

where $DVETH is the 'eth' interface that the OF controller is using. With this 50ms delay, the ZodiacFX start to report traces like the following, which indicate it is no processing some of the OF messages sent from the OF controller:

```
openflow.c: Processing 96 byte OpenFlow message 87085161 (488)
openflow_13.c: 87085161: OpenFlow message received type = 14
openflow_13.c: Allocating 56 bytes at 0x20003600 for flow mode in flow 13
openflow_13.c: Allocating 16 bytes at 0x20001420 for match field in flow 13
openflow_13.c: Allocating 24 bytes at 0x20001c20 for instruction field in flow 13
openflow_13.c: New flow added at 14 into table 3 : priority 12320 : cookie 0x5adc15c0
switch.c: 591 byte received from controller
openflow.c: OpenFlow data received (536 bytes)
switch.c: 591 byte received from controller
openflow.c: OpenFlow data received (536 bytes)
switch.c: 343 byte received from controller
openflow.c: OpenFlow data received (288 bytes)
switch.c: 303 byte received from controller
openflow.c: OpenFlow data received (248 bytes)
switch.c: 591 byte received from controller
openflow.c: OpenFlow data received (536 bytes)
switch.c: 591 byte received from controller
openflow.c: OpenFlow data received (536 bytes)
switch.c: 503 byte received from controller
openflow.c: OpenFlow data received (448 bytes)
openflow.c: Processing 96 byte OpenFlow message 87085190 (96)
openflow_13.c: 87085190: OpenFlow message received type = 14
openflow_13.c: Allocating 56 bytes at 0x20003638 for flow mode in flow 14
openflow_13.c: Allocating 6 bytes at 0x20001430 for match field in flow 14
openflow_13.c: Allocating 32 bytes at 0x20003670 for instruction field in flow 14
openflow_13.c: New flow added at 15 into table 1 : priority 4096 : cookie 0x5adc15c0
```

**Comment by pzanna** [comment]
This has now been fixed in v0.86
